tests/datasource/fluent/test_snowflake_datasource.py

```diff
@@ -205,15 +208,131 @@ def noop(self):
+    def test_fmt1_hyphen_parse(self, separator: Literal[".", "-"]):
+        orgname = "orgname"
+        account_name = "account-name"
+        value = f"{orgname}{separator}{account_name}"
+        print(f"{value=}")
+
+        account_identifier = pydantic.parse_obj_as(AccountIdentifier, value)
+        assert account_identifier.match
+
+        assert account_identifier.account_name == account_name
+        assert account_identifier.orgname == orgname
+        assert account_identifier.as_tuple() == (orgname, account_name)
+
+    @pytest.mark.parametrize("separator", [".", "-"])
+    def test_fmt1_underscore_parse(self, separator: Literal[".", "-"]):
+        orgname = "orgname"
+        account_name = "account_name"
+        value = f"{orgname}{separator}{account_name}"
+        print(f"{value=}")
+
+        account_identifier = pydantic.parse_obj_as(AccountIdentifier, value)
+        assert account_identifier.match
+
+        assert account_identifier.account_name == account_name
+        assert account_identifier.orgname == orgname
+        assert account_identifier.as_tuple() == (orgname, account_name)
+
+    @pytest.mark.parametrize(
+        "value",
+        [
+            "abc12345.us-east-1.aws",
+            "xy12345.us-gov-west-1.aws",
+            "xy12345.europe-west4.gcp",
+            "xy12345.central-us.azure",
+        ],
+    )
+    def test_fmt2_parse(self, value: str):
+        """The cloud portion is technically optional if the the provider is AWS, but expecting greatly simplifies our parsing logic"""
+        print(f"{value=}")
+        locator, _, _remainder = value.partition(".")
+        cloud_region_id, _, cloud_service = _remainder.partition(".")
+
+        account_identifier = pydantic.parse_obj_as(AccountIdentifier, value)
+        assert account_identifier.match
+
+        assert account_identifier.account_locator == locator
+        assert account_identifier.region == (cloud_region_id or None)
+        assert account_identifier.cloud == (cloud_service or None)
+
+        assert account_identifier.as_tuple() == (
+            locator,
+            cloud_region_id,
+            cloud_service,
+        )
+
+    @pytest.mark.parametrize(
+        "value",
+        [
+            "foobar",
+            "my_account.us-east-1",
+            "xy12345.us-gov-west-1.aws.",
+            "xy12345.europe-west4.gcp.bar",
+            "xy12345.us-east-1.nope",
+            "xy12345.",
+            "xy12345.central-us.bazar",
+            "xy12345_us-gov-west-1_aws",
+            "xy12345_europe-west4_gcp",
+            "xy12345_central-us_azure",
+        ],
+    )
+    def test_invalid_formats(self, value: str):
+        """Test that an invalid format does not match but can still be stringified as the original value."""
+        print(f"{value=}")
+        account_identifier = pydantic.parse_obj_as(AccountIdentifier, value)
+        assert not account_identifier.match
+        assert str(account_identifier) == value
```

Comment:
@DrewHoo @annabarr 
This is probably the most important part of this PR to review.
Specifically, the test cases around which values are/aren't considered to match the expected formats.